### ISSUE TITLE
[FR] autolevel fade height and extruder "output"

### ISSUE THREAD

**Issue by spilz87** [MarlinFirmware/Marlin#15053]
### Description

When I use fade height, Z move when X and Y move, which is normal, but Extruder put the same quantity of material what ever if fade "increases" or "reduces" the actual layer 

### Steps to Reproduce

use a non perfect leveled bed
add G29 and enable 1mm fade height  
-> where bed is lower, there is not enouth filament
-> where bed is higher, there is too much filament

I expect fade height change the extruder "output" according to Z compensation

regards

**Comment by AnHardt** [comment]
Set a larger fade height to make the difference between the layers smaller. Than the different worm shape (more or less squeezed) is able to bridge the vertical gaps (on cost of width). 

"Extrusion Amount Leveling" still isn't implemented in Marlin. Is it somewhere else?

**Comment by spilz87** [comment]
Thanks for your reply
It’s a way to compensate it 

Regards

**Comment by Roxy-3D** [comment]
1 mm of 'Fade Height' is not very much.   There is a reason 10 mm is the default amount.

With that said...   the math to adjust the extrusion amount is not very complicated.  I'm pretty sure we don't want to add any trigonometric calculations into the calculations.   But we can probably get very good approximations on how much to scale the number based on the length of the line segment and the difference in Z-Height correction.    

One simplifying assumption would be the nozzle is laying down a rectangular cross section.   Of course that is not true.   But with this assumption, the math gets a lot simpler.  (And after all...  That is what the slicers are trying to do with each back and forth pass of the nozzle on the part.)

@AnHardt   Do you agree that we could get a good approximation on the scaling factor based on the line segment length and the difference in Z-Height correction?

**Comment by AnHardt** [comment]
Never thought about that. Always planed to try to level, by extrusion amount, only at the first layer. But never found the time.

**Comment by AnHardt** [comment]
If you have the sliced extrusion width the additional filament volume is easily calculated from the difference between ideal and measured height. Together with the planed layer height a % or factor  value could be calculated.

**Comment by Roxy-3D** [comment]
Probably the biggest complication to adding the extrusion amount to the Z-Height correction is this:

The way the Z-Height correction is done, the desired Z height doesn't really change.   The firmware is just adding or subtracting a little bit to that amount.    That strategy doesn't work on the E-Axis because the machine is constantly pushing filament through the nozzle.   So, the firmware can't just decide to add a little bit more and bump the number up.   

Instead, we would have to schedule the E-Axis moves with a little bit more (or less) length and then correct the 'current position' to reflect where it would have been without any correction.

**Comment by AnHardt** [comment]
If we divide the fade-height by the layer-height we get the number of layers and fro that the amount of the above volume to correct per layer.

**Comment by Roxy-3D** [comment]
> If you have the sliced extrusion width the additional filament volume is easily calculated from the difference between ideal and measured height. Together with the planed layer height a % or factor value could be calculated.

Agreed...   But we don't really know the layer height.   However...  I don't think that matters.   The layer height is factored into what the slicer produces for the extrusion amount.   So if we just scale up (or down) what ever the GCode says to extrude for the line segment...   I think the layer height pretty much drops out of the equation.

**Comment by AnHardt** [comment]
How to get the numbers is a different theme and probably will be different if we are in `volumetric` mode or not.

**Comment by Roxy-3D** [comment]
Yes.   Good Point.

**Comment by boelle** [comment]
@spilz87 still having problems?

**Comment by spilz87** [comment]
It seems to change from a bug to a request but we can close as it’s not plane to add it

**Comment by boelle** [comment]
i have changed the title and added the FR label and will reopen, some day maybe
